packages/config/src/project/index.ts

```diff
@@ -0,0 +1,134 @@
+import _ from 'lodash'
+import Debug from 'debug'
+import deepDiff from 'return-deep-diff'
+
+import errors, { ConfigValidationFailureInfo, CypressError } from '@packages/errors'
+import { getCtx } from '@packages/data-context/src/globalContext'
+import type {
+  ResolvedFromConfig, TestingType, FullConfig,
+} from '@packages/types'
+
+import {
+  validate,
+  validateNoBreakingConfig,
+} from '../browser'
+import {
+  setPluginResolvedOn,
+  mergeDefaults,
+} from './utils'
+
+const debug = Debug('cypress:config:project')
+
+// TODO: any -> SetupFullConfigOptions in data-context/src/data/ProjectConfigManager.ts
+export function setupFullConfigWithDefaults (obj: any = {}): Promise<FullConfig> {
+  debug('setting config object %o', obj)
+  let { projectRoot, projectName, config, envFile, options, cliConfig } = obj
+
+  // just force config to be an object so we dont have to do as much
+  // work in our tests
+  if (config == null) {
+    config = {}
+  }
+
+  debug('config is %o', config)
+
+  // flatten the object's properties into the master config object
+  config.envFile = envFile
+  config.projectRoot = projectRoot
+  config.projectName = projectName
+
+  // @ts-ignore
+  return mergeDefaults(config, options, cliConfig)
```

Comment:
How come we need to ts-ignore here?

Reply:
It's because of the many errors inside and outside `mergeDefaults` like these:

![image](https://user-images.githubusercontent.com/8130013/179679580-0c3c5505-8af0-4005-9950-c793e084d8ab.png)

![image](https://user-images.githubusercontent.com/8130013/179679683-8244ea4d-d91c-4e72-b68c-17b3ea00a237.png)

I'll try to fix this in the subsequent PR.

Reply:
Alright, I figured it was something like this.